### PR TITLE
bpo-40334: Improvements to error-handling code in the PEG parser

### PR DESCRIPTION
**Pull request by lysnikolaou**
Following improvements are implemented in this PR:
- `p->error_indicator` is set, in case malloc or realloc fail
- Avoid memory leaks in the case that realloc fails
- Call `PyErr_NoMemory()` instead of `PyErr_Format()`, because it
  requires no memory

https://bugs.python.org/issue40334

### CODE REVIEW

**Comment by lysnikolaou** [comment]
Note to self: This partially conflicts with #19987, so will probably need to merge master after that's landed.

**Comment by pablogsal** [comment]
> * `p->error_indicator` is set, in case malloc or realloc fail

Do we need to do this? As soon as the rule returns it will go immediately to a `
if (_res == NULL && PyErr_Occurred())` block, no?

**Comment by lysnikolaou** [comment]
> > * `p->error_indicator` is set, in case malloc or realloc fail
> 
> Do we need to do this? As soon as the rule returns it will go immediately to a ` if (_res == NULL && PyErr_Occurred())` block, no?

I can't really see this, how exactly is this?

**Comment by pablogsal** [comment]
> I can't really see this, how exactly is this?

I have checked again and this only happens when the rule does not return `NULL`, so forget what I said :(

**Comment by lysnikolaou** [comment]
Resolved the conflicts so that we can hopefully merge this and close bpo-40334.

File: Tools/peg_generator/pegen/c_generator.py [hunk 1]
```diff
@@ -682,10 +674,12 @@ def handle_alt_loop(self, node: Alt, is_gather: bool, rulename: Optional[str]) -
...
-                self.print(
-                    "_children = PyMem_Realloc(_children, _children_capacity*sizeof(void *));"
-                )
-                self.out_of_memory_return(f"!_children", "NULL", message=f"realloc {rulename}")
+                self.print("void **_new_children = PyMem_Realloc(_children, _children_capacity*sizeof(void *));")
+                self.out_of_memory_return(f"!_new_children")
+                self.print("else {") # out_of_memory_return generates an if
+                with self.indent():
+                    self.print("_children = _new_children;")
+                self.print("}")
```
Comment (pablogsal) [inline Tools/peg_generator/pegen/c_generator.py]: I think we can eliminate this `else`:
```suggestion
                self.print("_children = _new_children;")
```

Comment (lysnikolaou) [reply Tools/peg_generator/pegen/c_generator.py]: Yup.

**Comment by pablogsal** [comment]
@lysnikolaou You need to regenerate the parser 😉

**Comment by lysnikolaou** [comment]
> @lysnikolaou You need to regenerate the parser 😉

Done! It's getting late..

**Comment by pablogsal** [comment]
> Done! Sorry, it's getting late..

No problem. I want to land this and the other PR so they both make it to beta 1 which is next Monday :)

**Pull request merged**